Add Elm

openorclose opened the pull request:
Fixes #145

openorclose committed:
Add Elm

damithc reviewed (commented):
After a quick glance, feels like this chapter is going too deep into details of Elm. Let's not stray into _tutorial_ territory (we are not trying to teach Elm to the reader). Technical details such as architecture may not be of interest to our target reader either, unless they are needed to explain a benefit or a key characteristic of Elm, in which case keep these details as short as possible.
Just like in lightning talks, figure out your key points and give just enough information to support those key points.

openorclose commented:
> After a quick glance, feels like this chapter is going too deep into details of Elm. Let's not stray into _tutorial_ territory (we are not trying to teach Elm to the reader). Technical details such as architecture may not be of interest to our target reader either, unless they are needed to explain a benefit or a key characteristic of Elm, in which case keep these details as short as possible.
> Just like in lightning talks, figure out your key points and give just enough information to support those key points.

Would it be better if I moved it to the frameworks section instead then?

Because Elm has no other uses other than for building frontend apps

damithc commented:
> Would it be better if I moved it to the frameworks section instead then?
> 
> Because Elm has no other uses other than for building frontend apps

Sure, if you think it is a framework rather than a language. My comment above applies regardless though.

openorclose committed:
Add Elm

openorclose commented:
Ok, I've removed the framework part of it.

alyip98 commented:
There's a part missing in the "What is Elm" section, I assume the missing part is web apps?

The "Why" section is well written and persuasive.

I find the "Basic Syntax" section to be rather dry and looks like yet another functional programming tutorial. I suggest shortening this section as it does not add much value.

If your aim is to show how it can be used to build web apps, perhaps a code example of a simple frontend app with a high level explanation would work well in illustrating how easy/useful/safe Elm is.

Just my 2 sen

openorclose committed:
Add TEA

openorclose commented:
> There's a part missing in the "What is Elm" section, I assume the missing part is web apps?
> 
> The "Why" section is well written and persuasive.
> 
> I find the "Basic Syntax" section to be rather dry and looks like yet another functional programming tutorial. I suggest shortening this section as it does not add much value.
> 
> If your aim is to show how it can be used to build web apps, perhaps a code example of a simple frontend app with a high level explanation would work well in illustrating how easy/useful/safe Elm is.
> 
> Just my 2 sen

Thanks for your suggestions :rocket: 

I've removed the basic syntax part and added in a brief overview of the Elm architecture.

openorclose committed:
Finish sentence

jamessspanggg commented on contents/elm/Elm.md:
```diff
@@ -0,0 +1,242 @@
+<div class="website-content">
+
+{{ booktitle | safe }}
+
```
perhaps can add a table of contents here? there's existing PRs adding TOC to all chapters

openorclose replied:
Added!

jamessspanggg commented on contents/elm/Elm.md:
```diff
@@ -0,0 +1,242 @@
+
+JavaScript, the standard language for web development, has a rather low learning curve but is considerably more difficult to maintain. To solve this, people write tests, add type annotations to JavaScript, and even try to adopt functional programming patterns like immutability to increase maintanability.
+
+Elm tries to be the best of both worlds. Its functional core provides maintainability, and it tries to greatly reduce the barriers to learning functional programming by simplifying concepts and having a really pleasant compiler.
```
...to learn

openorclose replied:
I think it's fine

jamessspanggg commented on contents/elm/Elm.md:
```diff
@@ -0,0 +1,242 @@
+
+### Benefit: Friendly Compiler
+
+Elm's compiler strives to be as friendly as possible, and offers as many hints as she can to help you fix errors. If nothing else, you should try Elm just to see how great error messages can be.
```
might not be good to assume the gender, haha. Perhaps can standardise to use `it` instead? as `it` is used in the What is Elm section

jamessspanggg commented on contents/elm/Elm.md:
```diff
@@ -0,0 +1,242 @@
+    └─────┘
+
+Learn more about why this is disallowed and how to break cycles
+here:<https://elm-lang.org/0.19.0/import-cycles>
```
Is is possible to bring out the link so that the user can click straight away?

jamessspanggg replied:
oh or is this a part of the error message?

openorclose replied:
Yup it's part of the error message

jamessspanggg commented on contents/elm/Elm.md:
```diff
@@ -0,0 +1,242 @@
+- a view function that takes in a state, and returns a HTML representation of that state
+- an update function that takes in a state and a message. The function then can use information provided in the message to returns a new state.
+
+Here's a minimal example (anything after `--` is a comment) :
```
may not be necessary to say that anything after `--` is a comment?

jamessspanggg reviewed (commented):
Good work! Very easy to read and understand from a beginner's standpoint. Some nits

openorclose committed:
Apply suggestions

openorclose commented:
> Good work! Very easy to read and understand from a beginner's standpoint. Some nits

Thanks for the review!

Tejas2805 commented:
Looks like you have followed [titleCase](https://apastyle.apa.org/style-grammar-guidelines/capitalization/title-case) in most places. Can you just go through and confirm once?

jamessspanggg commented on contents/elm/Elm.md:
```diff
@@ -0,0 +1,256 @@
+### Benefit: Handling All Cases
+Like many functional programming languages, you get the "if it compiles it works" feeling in Elm.
+
+If your code can produce more than one result (e.g. converting a string in to an integer can either fail or succeed), Elm's compiler forces you to account for all cases in order to compile.
```
in to -> into/to?

jamessspanggg commented on contents/elm/Elm.md:
```diff
@@ -0,0 +1,256 @@
+
+You can share your package, or use any of the packages available at https://package.elm-lang.org/.
+
+All packages there are written in Elm and therefore are safe (since Elm code cannot produce any side effects).
```
can put a tool tip on safe explaining (since Elm code cannot...)

openorclose replied:
I removed the word safe, and described it instead.

jamessspanggg commented on contents/elm/Elm.md:
```diff
@@ -0,0 +1,256 @@
+
+### Benefit: Fast Compiles and Small Bundles
+
+Elm's compiler is really fast. Its creator focused on compile times in its last two releases (0.19 and 0.19.1) which brought compile times for 100k LOC projects down to only 10s. If your project has less than 10k LOC, compilation should take no more than 1s.
```
it may not be the last two releases as time goes by

jamessspanggg reviewed (commented):
just some more nits

openorclose committed:
Apply suggestions

damithc commented on contents/elm/Elm.md:
````diff
@@ -0,0 +1,256 @@
+
+Learn more about why this is disallowed and how to break cycles
+here:<https://elm-lang.org/0.19.0/import-cycles>
+```
````
Should these output examples have line numbers?
On a related note, should we make line numbering default only if the code black has a language specifier? Is that even possible?

openorclose replied:
> Should these output examples have line numbers?
> On a related note, should we make line numbering default only if the code black has a language specifier? Is that even possible?

I'll hide them, don't think they're useful here.

I thought about that, but it would be rather confusing. Sometimes code blocks have line numbers and sometimes they do not depending on whether there is a langauge.

damithc replied:
> I thought about that, but it would be rather confusing. Sometimes code blocks have line numbers and sometimes they do not depending on whether there is a langauge.

Well, it's not entirely random and there is a good rationale. Might was users some work. It's nice when things 'just work' without the user having to do some additional configurations.

damithc replied:
Of course, if we do that, we also have to add a way to 'enable' line numbers too. So, that's another thing to consider on the cost side.

damithc commented on contents/elm/Elm.md:
```diff
@@ -0,0 +1,256 @@
+
+Examples of LOC vs compile times [here](https://discourse.elm-lang.org/t/how-are-0-19-compile-times-for-50k-loc-projects/3680).
+
+Lastly, Elm bundle size is small. Since Elm's functions are pure, tree-shaking can be done automatically. You can safely import a 1kLOC package only to use a one-liner function, and Elm can help you remove the 999 lines during compilation.
```
Add an explanatory note (e.g., popover) about tree-shaking?

damithc commented on contents/elm/Elm.md:
```diff
@@ -0,0 +1,256 @@
+
+## The Elm Architecture
+
+![](images/TEA.png)
```
Resize so that the text in the image matches the text size in the article?

damithc reviewed (commented):
Nice!
Mentioned a couple of nits.
Elm sounds like a wonderful language, which raises the question 'why is it not the most popular language'? Perhaps a brief note on its current state of adoption and also what it is not good for? I assume it is a JS replacement? Using Elm means giving up the ability to use all other popular farmeworks such as React, Angular, Vue?

openorclose committed:
Apply suggestions

openorclose commented:
> Nice!
> Mentioned a couple of nits.
> Elm sounds like a wonderful language, which raises the question 'why is it not the most popular language'? Perhaps a brief note on its current state of adoption and also what it is not good for? I assume it is a JS replacement? Using Elm means giving up the ability to use all other popular farmeworks such as React, Angular, Vue?

Thanks for the review prof!

I've added in some pain points of Elm, and how it can coexist together with other React or Vue code.

damithc reviewed (commented):
Seems almost ready now. You said you are going to hide line numbers of output text?

openorclose committed:
Remove line numbes from error messages

openorclose commented:
> Seems almost ready now. You said you are going to hide line numbers of output text?

Oops, missed that out.

I've hidden them now.
